src/model/ZoneHVACEvaporativeCoolerUnit.cpp

```diff
@@ -0,0 +1,530 @@
+      const bool result = setDouble(OS_ZoneHVAC_EvaporativeCoolerUnitFields::ShutOffRelativeHumidity, shutOffRelativeHumidity);
+      return result;
+    }
+
+    void ZoneHVACEvaporativeCoolerUnit_Impl::autosize() {
+      autosizeDesignSupplyAirFlowRate();
+    }
+
+    void ZoneHVACEvaporativeCoolerUnit_Impl::applySizingValues() {
+      if (boost::optional<double> val_ = autosizedDesignSupplyAirFlowRate()) {
+        setDesignSupplyAirFlowRate(*val_);
+      }
+    }
+
+    boost::optional<Schedule> ZoneHVACEvaporativeCoolerUnit_Impl::optionalAvailabilitySchedule() const {
+      return getObject<ModelObject>().getModelObjectTarget<Schedule>(OS_ZoneHVAC_EvaporativeCoolerUnitFields::AvailabilityScheduleName);
+    }
+
+    boost::optional<HVACComponent> ZoneHVACEvaporativeCoolerUnit_Impl::optionalSupplyAirFan() const {
+      return getObject<ModelObject>().getModelObjectTarget<HVACComponent>(OS_ZoneHVAC_EvaporativeCoolerUnitFields::SupplyAirFanName);
+    }
+
+    boost::optional<HVACComponent> ZoneHVACEvaporativeCoolerUnit_Impl::optionalFirstEvaporativeCooler() const {
+      return getObject<ModelObject>().getModelObjectTarget<HVACComponent>(OS_ZoneHVAC_EvaporativeCoolerUnitFields::FirstEvaporativeCooler);
+    }
+
+  }  // namespace detail
+
+  ZoneHVACEvaporativeCoolerUnit::ZoneHVACEvaporativeCoolerUnit(const Model& model)
+    : ZoneHVACComponent(ZoneHVACEvaporativeCoolerUnit::iddObjectType(), model) {
+    OS_ASSERT(getImpl<detail::ZoneHVACEvaporativeCoolerUnit_Impl>());
+
+    bool ok = true;
+    auto alwaysOn = model.alwaysOnDiscreteSchedule();
+    ok = setAvailabilitySchedule(alwaysOn);
+    OS_ASSERT(ok);
+
+    FanComponentModel supplyAirFan(model);
+    ok = setSupplyAirFan(supplyAirFan);
+    OS_ASSERT(ok);
+
+    autosizeDesignSupplyAirFlowRate();
+    ok = setFanPlacement("BlowThrough");
+    OS_ASSERT(ok);
+    ok = setCoolerUnitControlMethod("ZoneTemperatureDeadbandOnOffCycling");
+    OS_ASSERT(ok);
+    ok = setThrottlingRangeTemperatureDifference(1.0);
+    OS_ASSERT(ok);
+    ok = setCoolingLoadControlThresholdHeatTransferRate(100.0);
+    OS_ASSERT(ok);
+
+    EvaporativeCoolerDirectResearchSpecial firstEvaporativeCooler(model, alwaysOn);
+    ok = setFirstEvaporativeCooler(firstEvaporativeCooler);
+    OS_ASSERT(ok);
+
+    ok = setShutOffRelativeHumidity(100.0);
+    OS_ASSERT(ok);
+  }
+
+  ZoneHVACEvaporativeCoolerUnit::ZoneHVACEvaporativeCoolerUnit(const Model& model, Schedule& availabilitySchedule, HVACComponent& supplyAirFan,
+                                                               HVACComponent& firstEvaporativeCooler)
+    : ZoneHVACComponent(ZoneHVACEvaporativeCoolerUnit::iddObjectType(), model) {
+    OS_ASSERT(getImpl<detail::ZoneHVACEvaporativeCoolerUnit_Impl>());
+
+    bool ok = true;
+    ok = setAvailabilitySchedule(availabilitySchedule);
+    OS_ASSERT(ok);
+    ok = setSupplyAirFan(supplyAirFan);
+    OS_ASSERT(ok);
```

Comment:
Technically these can fail, and should be caught with a remove() because LOG_AND_THROW. Especially the supplyAirFan. I could pass any children of HVACComponent, such as a CoilCoolingDXSingleSpeed.